Split account types

tomsmalley opened the pull request.

Commit:
WIP Key Details dialog

tomsmalley committed:
Split account types: it builds

tomsmalley committed:
Fix some warnings

tomsmalley committed:
Fix account name validity check

tomsmalley committed:
Wire up most wallet events

alexfmpe commented on common/src/Common/Wallet.hs:
```diff
@@ -274,54 +302,239 @@ instance FromJSON UnfinishedCrossChainTransfer where
+    pure $ ref ==> nv
+
+accountUnfinishedCrossChainTransfer :: Account -> Maybe UnfinishedCrossChainTransfer
+accountUnfinishedCrossChainTransfer (r :=> Identity a) = _accountInfo_unfinishedCrossChainTransfer $ case r of
```
```suggestion
accountUnfinishedCrossChainTransfer = _accountInfo_unfinishedCrossChainTransfer . accountInfo
```

alexfmpe commented on common/src/Common/Wallet.hs:
```diff
@@ -274,54 +302,239 @@ instance FromJSON UnfinishedCrossChainTransfer where
+      }
+
+data VanityAccount = VanityAccount
+  { _vanityAccount_key :: PublicKey -- TODO should be KeySet, but that's a huge change.
```
Sounds like it won't be addressed by this PR? Maybe make a ticket?

tomsmalley replied:
We do have a ticket, but this is implicit. I'll make another.

alexfmpe commented on frontend/src/Frontend/Repl.hs:
```diff
@@ -308,9 +305,9 @@ setEnvKeys :: [KeyPair key] -> PactRepl (Term Name)
+      renderKeys = T.unwords . map (keyToText' . _keyPair_publicKey)
 
-      keyToText = safeDecodeUtf8 . BSL.toStrict . encode
+      keyToText' = safeDecodeUtf8 . BSL.toStrict . encode
```
Eh?

tomsmalley replied:
Name shadowing. I wasn't sure if it was different to the other one, possibly could be removed.

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -396,14 +390,16 @@ uiDeploymentSettings m settings = mdo
             mNetworkId = (hush . mkNetworkName . nodeVersion =<<) <$> currentNode
 
-            mHeadAccount = fmap _account_name . findFirstVanityAccount <$> (m ^. wallet_accounts)
+            accounts = liftA2 (\n -> Map.findWithDefault mempty n . unAccountStorage) (m ^. network_selectedNetwork) (m ^. wallet_accounts)
```
```suggestion
            accounts = liftA2 (Map.findWithDefault mempty) (m ^. network_selectedNetwork) (unAccountStorage <$> m ^. wallet_accounts)
```

alexfmpe commented on frontend/src/Frontend/UI/DeploymentSettings.hs:
```diff
@@ -1049,8 +1045,12 @@ uiDeployPreview model settings accounts signers gasLimit ttl code lastPublicMeta
+            Map.singleton (Some $ AccountRef_Vanity n c) (_vanityAccount_key va)
+        nonVanityAccounts = flip Map.foldMapWithKey (_accounts_nonVanity accounts) $ \pk ->
+          Map.foldMapWithKey $ \c _ ->
+            Map.singleton (Some $ AccountRef_NonVanity pk c) pk
```
How about defining, in `Common/Wallet.hs`
```haskell
foldAccounts :: Monoid m => (Account -> m) -> Accounts -> m
foldAccounts f accounts = mconcat
  [ flip Map.foldMapWithKey (_accounts_vanity accounts) $ \n -> Map.foldMapWithKey $ \c a -> f (AccountRef_Vanity n c ==> a)
  , flip Map.foldMapWithKey (_accounts_nonVanity accounts) $ \pk -> Map.foldMapWithKey $ \c a -> f (AccountRef_NonVanity pk c ==> a)
  ]
```
then this could be
```haskell
getAccounts = Map.mapKeys (\(Some x) -> AccountName $ accountRefToName x) . Map.restrictKeys accs
  where
    accs = flip foldAccounts accounts $ \case
      sum@(ref :=> _) -> Map.singleton (Some ref) $ accountKey sum
```

(I'm somewhat surprised I couldn't find a `DSum tag f -> Some tag` function)

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AccountDetails.hs:
```diff
@@ -74,17 +79,19 @@ uiAccountDetailsDetails key a onClose = Workflow $ do
+      _ <- displayText "Public Key" (keyToText $ accountKey a) "account-details__pubkey"
       -- Chain id
-      _ <- displayText "Chain ID" (Pact._chainId $ _account_chainId a) "account-details__chain-id"
+      _ <- displayText "Chain ID" (Pact._chainId $ accountChain a) "account-details__chain-id"
```
I'd move the `chain` binding further below to the `let` block before this, to use it here

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -176,7 +176,7 @@ uiReceiveModal0 model account onClose = Workflow $ do
             transactionDisplayNetwork model
             -- Chain id
-            _ <- displayText "Chain ID" (_chainId $ _account_chainId account) "account-details__chain-id"
+            _ <- displayText "Chain ID" (_chainId $ accountChain account) "account-details__chain-id"
```
`chain` is `let`-bound above

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Send.hs:
```diff
@@ -313,8 +327,10 @@ sendConfig model fromIndex fromAccount = Workflow $ do
+              pure $ ffor3 (model ^. wallet_accounts) (model ^. network_selectedNetwork) gasAcc $ \netToAccount net ma -> do
+                accounts <- Map.lookup net $ unAccountStorage netToAccount
+                a <- ma
+                lookupAccountRef a accounts
```
Might want to make this block be a `uiGasPayerDropdown` or so, since we do these exact lookups in the two other uses below.

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Send.hs:
```diff
@@ -672,17 +697,20 @@ continueCrossChainTransfer
         }
-      signingPairs = [_account_key gasPayer]
+      signingSet = Set.singleton $ accountKey gasPayer
+      signingPairs = filterKeyPairs signingSet keys
```
Can this ever not be `[_account_key gasPayer]` like before? Can it end up as `[]`?

tomsmalley replied:
If somehow the key has vanished it could be `[]`

alexfmpe commented on frontend/src/Frontend/UI/Wallet.hs:
```diff
@@ -147,19 +145,16 @@ uiKeyItems model = do
-      dyn_ $ ffor2 keyMap selectedNetwork $ \keys net ->
-        unless (any (isJust . activeAccountOnNetwork net) $ Map.elems keys) $
+      dyn_ $ ffor keyMap $ \keys ->
+        when (all _key_hidden $ Map.elems keys) $
```
Don't need `Map.elems`

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -119,66 +107,40 @@ makePactLenses ''WalletCfg
-getSigningPairs :: Set AccountName -> Accounts key -> [KeyPair key]
-getSigningPairs signing = fmapMaybe isForSigning . IntMap.elems
+getSigningPairs :: KeyStorage key -> Accounts -> Set (Some AccountRef) -> [KeyPair key]
+getSigningPairs allKeys allAccounts signing = Map.elems $ Map.restrictKeys allKeysMap wantedKeys
```
```haskell
getSigningPairs allKeys allAccounts signing = filterKeyPairs wantedKeys allKeys
```

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -119,66 +107,40 @@ makePactLenses ''WalletCfg
 -- | An empty wallet that will never contain any keys.
 emptyWallet :: Reflex t => Wallet key t
-emptyWallet = Wallet mempty never
+emptyWallet = mempty
```
Binding doesn't seem to actually be used anywhere besides the export

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -187,78 +149,70 @@ nextKey :: IntMap a -> Int
-    onNewKey <- performEvent $ attachWith (createKey . nextKey) (current keys) onGenKey
-    onWOAccountCreate <- performEvent $ attachWith (createWalletOnlyAccount . nextKey) (current keys) onCreateWOAcc
-    newBalances <- getBalances model $ current keys <@ refresh
+    onNewKey <- performEvent $ attachWith (\a _ -> createKey $ nextKey a) (current keys) onGenKey
```
```haskell
    onNewKey <- performEvent $ createKey . nextKey <$> current keys <@ onGenKey
```

alexfmpe commented on common/src/Common/Wallet.hs:
```diff
@@ -274,54 +302,239 @@ instance FromJSON UnfinishedCrossChainTransfer where
+      { _nonVanityAccount_info = info
+      }
+
+newtype AccountStorage = AccountStorage { unAccountStorage :: Map NetworkName Accounts }
```
Is there a reason (besides the code currently being extremely biased towards it, I mean) for `NetworkName` to be outside `Accounts`, but `ChainId` be inside?

tomsmalley replied:
Only the structure of the JSON and to reflect the split a little more

tomsmalley committed:
Update common/src/Common/Wallet.hs

Co-Authored-By: Alexandre Esteves <2335822+alexfmpe@users.noreply.github.com>

tomsmalley committed:
Update frontend/src/Frontend/UI/DeploymentSettings.hs

Co-Authored-By: Alexandre Esteves <2335822+alexfmpe@users.noreply.github.com>

tomsmalley committed:
Review changes

tomsmalley committed:
Merge branch 'ts-split-accounts' of github.com:kadena-io/chainweaver into ts-split-accounts

tomsmalley committed:
Store accounts

tomsmalley committed:
Merge remote-tracking branch 'origin/develop' into ts-split-accounts

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -129,78 +113,34 @@ makePactLenses ''WalletCfg
+findFirstVanityAccount as = do
+  (n, cm) <- Map.lookupMin $ _accounts_vanity as
+  (c, va) <- Map.lookupMin cm
+  pure (n, c, va)
```
Seems like these are both used only once. How about instead defining only
```haskell
findFirstVanityAccount pred as = do
  (n, cm) <-  Map.lookupMin $ _accounts_vanity as
  (c, va) <- Map.lookupMin $ Map.filter pred cm
  pure (n, c, va)
```
and calling it as `findFirstVanityAccount _vanityAccount_inflight` or `findFirstVanityAccount (const True)`

alexfmpe committed:
Separate account and key routes

alexfmpe commented on frontend/src/Frontend/UI/Wallet.hs:
```diff
@@ -146,19 +144,16 @@ uiKeyItems model = do
+      dyn_ $ ffor keyMap $ \keys ->
+        when (all _key_hidden keys) $
           elClass "tr" "wallet__table-row" $ elAttr "td" ("colspan" =: "6" <> "class" =: "wallet__table-cell") $
             text "No accounts ..."
```
"No keys ..."

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AccountDetails.hs:
```diff
@@ -35,14 +37,15 @@ uiAccountDetails
-  => IntMap.Key
-  -> Account key
+  => NetworkName
+  -> AccountName
```
Can we obtain this by calling `accountToName` on the `Account` argument?

mankyKitty replied:
Yep, also the account details dialog is being upgraded to work on both vanity & non-vanity accounts, so this will likely change with my PR.

alexfmpe committed:
Populate accounts table

alexfmpe committed:
Fix keys icon

Commit:
Allow account details dialog to accept either kind of account

Commit:
Reuse function from the wallet, move var out to let

alexfmpe committed:
Fix columns of keys table

alexfmpe committed:
Update details button

alexfmpe committed:
Factor out accordion button

alexfmpe committed:
Stub account rows in keys table

alexfmpe committed:
Fix button background

alexfmpe committed:
Fix colspan

alexfmpe committed:
Merge branch 'develop' into ts-split-accounts

Commit:
Tighten up KeyDetails dialog for new types

Commit:
Updated details with the notes for keys

tomsmalley committed:
Add someTag utility function

tomsmalley committed:
Accounts plumbing

Commit:
Fixes from PR feedback and added some lenses for Account and AccountInfo

Commit:
Giving it back

tomsmalley committed:
Remove AddAccount modal, use AddVanityAccount modal directly

tomsmalley committed:
Merge remote-tracking branch 'origin/sc-key-details-dialog' into ts-split-accounts

tomsmalley committed:
Merge remote-tracking branch 'origin/sc-generalise-details-vanity-nonvanity' into ts-split-accounts

tomsmalley committed:
Account details modal

tomsmalley committed:
Tweak wording for keys UI

tomsmalley committed:
Add missing deps to common

tomsmalley committed:
Add missing RankNTypes pragma

tomsmalley committed:
Add missing frontend dependency

tomsmalley committed:
De-dupe similar functions

tomsmalley committed:
Merge remote-tracking branch 'origin/develop' into ts-split-accounts

alexfmpe commented on frontend/src/Frontend/UI/Wallet.hs:
```diff
@@ -124,101 +229,139 @@ uiKeyItems
+  switchHold never <=< dyn $ ffor hidden $ \case
+    True -> pure never
+    False -> mdo
+      rec
```
`rec` inside `mdo` ?

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AddVanityAccount.hs:
```diff
@@ -65,13 +59,40 @@ type HasUISigningModelCfg mConf key t =
+  -> m mConf
+uiAddVanityAccountButton m = do
+  eOpenAddAccount <- uiButton (def & uiButtonCfg_class <>~ " main-header__add-account-button")  (text "+ Add Account")
+  pure $ mempty & modalCfg_setModal .~ (Just (uiAddVanityAccountDialog m) <$ eOpenAddAccount)
```
`?~`

tomsmalley replied:
It's `Event (Maybe x)`

alexfmpe replied:
Oops, misread

tomsmalley committed:
Fix nested recursive block

tomsmalley committed:
Make key notes updatable

tomsmalley committed:
Merge remote-tracking branch 'origin/develop' into ts-split-accounts

tomsmalley committed:
Fix name conflicts

tomsmalley committed:
Add questionable hiding functionality

tomsmalley committed:
Merge remote-tracking branch 'origin/develop' into ts-split-accounts

benkolera committed:
feat: Refactor storage to a free monad for great testability

benkolera committed:
Keys refactor, storage backup

benkolera committed:
Upgrade from V0 to V1 done. Needs some tests.

benkolera committed:
refactor: Split out in memory storage so that can reuse for migration tests

benkolera committed:
feat: Migration looks pretty good. Just need some better test data.

benkolera committed:
feat: Migration tested and works in devDesktop

benkolera committed:
fix: Parses old public keys correctly

benkolera committed:
fix: Requests all chains for key balances.

Commit:
Change unknown balance msg

Change the text for unknown balance information from "Unknown" to "Account not present"

benkolera committed:
Remove traceevent

Commit:
Move 'Receive' from top level to individual keys.

This moves the button for the Receive dialog to the individual keys on each chain, pinning the
respective chainId in the dialog. The 'created' status is selected based on the presence of
the account balance for that key.

Commit:
Remove reflexive style-based refactor

Commit:
Clean up KeyDialog type a bit for simpler use

tomsmalley committed:
Add missing deps and pragmas

tomsmalley committed:
Alter StorageVersioner naming

Commit:
Show full public key on Keys page

tomsmalley committed:
Make V0 types match released types

tomsmalley committed:
'maybe mempty id' -> 'fold'

benkolera committed:
Fixed up test to remove inflight object in test data

tomsmalley committed:
MTL style storage

tomsmalley committed:
Fix build

tomsmalley committed:
Remove old goldentests

tomsmalley committed:
Deduplicate aeson encoding

Commit:
Remove the chain table from the key details dialog. Fix an issue where the horizontal separator would display incorrectly on account details dialog

tomsmalley committed:
Fix file storage file names

Commit:
Move comment to be with its associated code

benkolera committed:
Merge branch 'ts-split-accounts' into blk@ts-split-accounts/versioning

tomsmalley committed:
Merge pull request #368 from kadena-io/blk@ts-split-accounts/versioning

Schema Migrations

tomsmalley committed:
Merge pull request #369 from kadena-io/sc-change-unknown-to-accnotpresent

Change unknown balance display text

tomsmalley committed:
Merge pull request #371 from kadena-io/sc-remove-ellipsis-keyspage

Show full public key on Keys page

tomsmalley committed:
Merge remote-tracking branch 'origin/ts-split-accounts' into sc-move-receive-to-nonvanity

tomsmalley committed:
Merge pull request #370 from kadena-io/sc-move-receive-to-nonvanity

Move 'Receive' from top level to individual keys

tomsmalley committed:
Merge remote-tracking branch 'origin/ts-split-accounts' into sc-remove-chaintable

tomsmalley committed:
Merge pull request #372 from kadena-io/sc-remove-chaintable

Remove the chain table from the key details dialog.

alexfmpe committed:
Show KDA units in headers and not in every row

tomsmalley committed:
Merge pull request #373 from kadena-io/ae@kda-header

Show KDA units in headers and not in every row

tomsmalley committed:
Add missing test modules to cabal file
